crates/ty_python_semantic/resources/mdtest/generics/legacy/classes.md

```diff
@@ -329,6 +329,57 @@ class D(C[V, int]):
 reveal_type(D(1))  # revealed: D[int]
 ```
 
+### Generic class inherits `__init__` from generic base class
+
+```py
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+class C(Generic[T, U]):
+    def __init__(self, t: T, u: U) -> None: ...
+
+class D(C[T, U]):
+    pass
+
+reveal_type(C(1, "str"))  # revealed: C[int, str]
+reveal_type(D(1, "str"))  # revealed: D[int, str]
+```
+
+### Generic class inherits `__init__` from `dict`
+
+This is a specific example of the above, since it was reported specifically by a user.
+
+```py
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+class D(dict[T, U]):
+    pass
+
+reveal_type(D(key=1))  # revealed: D[str, int]
+```
+
+### Generic class inherits `__new__` from `tuple`
+
+(Technically, we synthesize a `__new__` method that is more precise than the one defined in typeshed
+for `tuple`, so we use a different mechanism to make sure it has the right inherited generic
+context. But from the user's point of view, this is another example of the above.)
+
+```py
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+class C(tuple[T, U]): ...
+
+reveal_type(C((1, 2)))  # revealed: C[int, int]
```

Comment:
I sort of wish that this was `C[Literal[1], Literal[2]]` -- I don't think there's a strong reason to upcast the `Literal` types to `int` types here, since `tuple` is covariant. We'd obviously infer `tuple[Literal[1], Literal[2]]` rather than `tuple[int, int]` for `(1, 2)`, for similar reasons: the covariance of `tuple` means that `tuple[Literal[1], Literal[2]]` is assignable to `tuple[int, int]`, so there's no way that inferring the more precise type can produce false positives.

But that's not for this PR! It's obviously a pre-existing issue.

Reply:
Yep totally!